File: typescript/packages/core/src/core/hierarchy/readdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { Accessor } from '../../accessor/base.ts'
import type { IndexEntry } from '../../cache/index/config.ts'
import { RAMIndexCacheStore } from '../../cache/index/ram.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { PathSpec } from '../../types.ts'
import { enoent, enotdir } from '../../utils/errors.ts'
import { mountKey, mountPrefixOf } from '../../utils/key_prefix.ts'
import { rstripSlash, stripSlash } from '../../utils/slash.ts'
import { compareCodePoints } from '../../utils/sort.ts'
import { resolveEntry, type ReaddirFn } from './probe.ts'
import { INVALID, ROOT, type DetectFn, type ScopeMatch } from './scope.ts'

export type Lister<A extends Accessor> = (
  accessor: A,
  match: ScopeMatch,
) => Promise<[string, IndexEntry][] | null>

export type EntryLister<A extends Accessor> = (
  accessor: A,
  match: ScopeMatch,
  entry: IndexEntry,
) => Promise<[string, IndexEntry][]>

export type Guard<A extends Accessor> = (
  accessor: A,
  match: ScopeMatch,
  virtual: string,
) => Promise<void>

/**
 * Build a hierarchy readdir: dispatch, guards, index, name joins.
 *
 * A lister fetches one directory kind and returns `[vfsName, IndexEntry]`
 * pairs; everything else — classification, existence guards, the index probe
 * and write-back, and virtual name construction — happens here, identically
 * for every backend. A dot-prefixed name is dropped from the listing: the
 * classifier refuses every dot-leading segment, so listing one would
 * advertise a path that stat, read and child readdir all report absent.
 * A lister may answer null instead of a listing: the directory's container
 * does not exist, reported as ENOENT on the virtual path.
 *
 * An entry lister is for a directory whose existence and contents are
 * already proven by its parent's listing: the kit resolves the directory's
 * own index entry through `resolveEntry` (warming parent listings, each one
 * cached) and hands it over, so entering a directory a traversal just listed
 * costs no API call at all. A container lister that instead re-fetched its
 * ancestor chain per directory made a recursive walk quadratic in listing
 * payloads. The facts a child listing needs beyond the API's own answers
 * ride the parent listing's `IndexEntry.extra` (trello stashes each
 * `card.json` size on the card's directory entry).
 *
 * `listers` holds one lister per directory kind; include `root` for a
 * dynamic mount root. `entryListers` holds listers for kinds resolved
 * through their parent's listing; a kind appears in exactly one of the two
 * tables. `staticRoot` names fixed top-level entries, for backends whose
 * root never changes; it bypasses the index. `guards` are existence checks
 * that run before the index probe, so a vanished container is ENOENT even on
 * a warm cache. `leafError` is what listing a leaf raises; fixed hierarchies
 * historically answer ENOENT.
 */
export function makeReaddir<A extends Accessor>(
  detect: DetectFn,
  options: {
    listers: Readonly<Record<string, Lister<A>>>
    entryListers?: Readonly<Record<string, EntryLister<A>>>
    staticRoot?: readonly string[]
    guards?: Readonly<Record<string, Guard<A>>>
    leafError?: 'enoent' | 'enotdir'
  },
): ReaddirFn<A> {
  const { listers, staticRoot, guards } = options
  const entryListers = options.entryListers ?? {}
  const leafError = options.leafError ?? 'enoent'
  const overlap = Object.keys(listers).filter((kind) => entryListers[kind] !== undefined)
  if (overlap.length > 0) {
    throw new Error(`kinds in both lister tables: ${overlap.sort(compareCodePoints).join(', ')}`)
  }
  return async function readdir(
    accessor: A,
    pathSpec: PathSpec,
    index?: IndexCacheStore,
  ): Promise<string[]> {
    // Entry resolution and the parent-listing warm both read what readdir
    // just wrote, so a caller with no cache still needs one for the
    // duration of the call.
    const store = index ?? new RAMIndexCacheStore()
    const virtual = pathSpec.virtual
    const prefix = mountPrefixOf(pathSpec.virtual, pathSpec.resourcePath)
    const path = (pathSpec.pattern !== null ? pathSpec.dir : pathSpec).mountPath
    const key = stripSlash(path)
    const virtualKey = key !== '' ? `${prefix}/${key}` : prefix !== '' ? prefix : '/'
    const match = detect(path)
    if (match.kind === INVALID) throw enoent(pathSpec)
    if (match.kind === ROOT && staticRoot !== undefined) {
      return staticRoot.map((d) => `${prefix}/${d}`)
    }
    const lister = listers[match.kind]
    const entryLister = entryListers[match.kind]
    if (lister === undefined && entryLister === undefined) {
      if (match.scope !== null && match.scope.leaf && leafError === 'enotdir') {
        throw enotdir(pathSpec)
      }
      throw enoent(pathSpec)
    }
    const guard = guards?.[match.kind]
    if (guard !== undefined) await guard(accessor, match, virtual)
    const listing = await store.listDir(virtualKey)
    if (listing.entries !== undefined && listing.entries !== null) return listing.entries
    let listed: [string, IndexEntry][]
    if (entryLister !== undefined) {
      const own = await resolveEntry(
        readdir,
        accessor,
        new PathSpec({
          virtual: virtualKey,
          directory: virtualKey,
          resolved: false,
          resourcePath: mountKey(virtualKey, prefix),
        }),
        store,
      )
      if (own === null) throw enoent(pathSpec)
      listed = await entryLister(accessor, match, own)
    } else if (lister !== undefined) {
      const maybe = await lister(accessor, match)
      if (maybe === null) throw enoent(pathSpec)
      listed = maybe
    } else {
      throw enoent(pathSpec)
    }
    const entries = listed.filter(([name]) => !name.startsWith('.'))
    await store.setDir(virtualKey, entries)
    const stem = rstripSlash(virtualKey)
    return entries.map(([name]) => `${stem}/${name}`)
  }
}

File: python/mirage/core/hierarchy/readdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import Awaitable, Callable, Mapping
from typing import Literal

from mirage.cache.index import NULL_INDEX, IndexCacheStore, IndexEntry
from mirage.core.hierarchy.probe import A, ReaddirFn
from mirage.core.hierarchy.scope import INVALID, ROOT, DetectFn, RouteMatch
from mirage.types import PathSpec
from mirage.utils.errors import enoent, enotdir
from mirage.utils.key_prefix import mount_prefix_of

Lister = Callable[[A, RouteMatch], Awaitable[list[tuple[str, IndexEntry]]]]
Guard = Callable[[A, RouteMatch, str], Awaitable[None]]


def make_readdir(
    detect: DetectFn,
    *,
    listers: Mapping[str, Lister[A]],
    static_root: tuple[str, ...] | None = None,
    guards: Mapping[str, Guard[A]] | None = None,
    leaf_error: Literal["enoent", "enotdir"] = "enoent",
) -> ReaddirFn[A]:
    """Build a hierarchy readdir: dispatch, guards, index, name joins.

    A lister fetches one directory kind and returns ``(vfs_name,
    IndexEntry)`` pairs; everything else — classification, existence
    guards, the index probe and write-back, and virtual name
    construction — happens here, identically for every backend.

    Args:
        detect (DetectFn): the backend's route classifier.
        listers (Mapping[str, Lister]): one lister per directory kind;
            include ``root`` for a dynamic mount root.
        static_root (tuple[str, ...] | None): fixed top-level names, for
            backends whose root never changes; bypasses the index.
        guards (Mapping[str, Guard]): existence checks that run before
            the index probe, so a vanished container is ENOENT even on a
            warm cache.
        leaf_error (Literal["enoent", "enotdir"]): what listing a leaf
            raises; fixed hierarchies historically answer ENOENT.
    """

    async def readdir(accessor: A,
                      path_spec: PathSpec,
                      index: IndexCacheStore = NULL_INDEX) -> list[str]:
        virtual = path_spec.virtual
        prefix = mount_prefix_of(path_spec.virtual, path_spec.resource_path)
        path = (path_spec.dir if path_spec.pattern else path_spec).mount_path
        key = path.strip("/")
        virtual_key = prefix + "/" + key if key else prefix or "/"
        match = detect(path)
        if match.kind == INVALID:
            raise enoent(virtual)
        if match.kind == ROOT and static_root is not None:
            return [f"{prefix}/{d}" for d in static_root]
        lister = listers.get(match.kind)
        if lister is None:
            if (match.route is not None and match.route.leaf
                    and leaf_error == "enotdir"):
                raise enotdir(virtual)
            raise enoent(virtual)
        guard = guards.get(match.kind) if guards is not None else None
        if guard is not None:
            await guard(accessor, match, virtual)
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
        entries = await lister(accessor, match)
        await index.set_dir(virtual_key, entries)
        stem = virtual_key.rstrip("/")
        return [f"{stem}/{name}" for name, _ in entries]

    return readdir
